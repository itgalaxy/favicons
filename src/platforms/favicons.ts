import { FaviconHtmlElement } from "../index";
import { FaviconOptions, NamedIconOptions } from "../config/defaults";
import { transparentIcon, transparentIcons } from "../config/icons";
import { OptionalMixin, Platform, uniformIconOptions } from "./base";

const ICONS_OPTIONS: (NamedIconOptions & OptionalMixin)[] = [
  { name: "favicon.ico", ...transparentIcons(16, 24, 32, 48, 64) },
  { name: "favicon-16x16.png", ...transparentIcon(16) },
  { name: "favicon-32x32.png", ...transparentIcon(32) },
  { name: "favicon-48x48.png", ...transparentIcon(48) },
  { name: "favicon.svg", ...transparentIcon(1024), optional: true }, // arbitrary size. if more than one svg source is given, the closest to this size will be picked.
];

export class FaviconsPlatform extends Platform {
  constructor(options: FaviconOptions) {
    super(
      options,
      uniformIconOptions(options, options.icons.favicons, ICONS_OPTIONS),
    );
  }

  override async createHtml(): Promise<FaviconHtmlElement[]> {
    return this.iconOptions.map(({ name, ...options }) => {
      if (name.endsWith(".ico")) {
        // prettier-ignore
        return `<link rel="icon" type="image/x-icon" href="${this.cacheBusting(this.relative(name))}">`;
      } else if (name.endsWith(".svg")) {
        // prettier-ignore
        return `<link rel="icon" type="image/svg+xml" href="${this.cacheBusting(this.relative(name))}">`;
      }

      const { width, height } = options.sizes[0];

      // prettier-ignore
      return `<link rel="icon" type="image/png" sizes="${width}x${height}" href="${this.cacheBusting(this.relative(name))}">`;
    });
  }
}
