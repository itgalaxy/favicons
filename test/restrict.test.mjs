import favicons from "../src";
import { logo_png } from "./util";

test("should allow to restrict the icons to generate a favicon only", async () => {
  expect.assertions(2);
  const { html, images } = await favicons(logo_png, {
    icons: {
      favicons: ["favicon-32x32.png"],
      android: false,
      appleIcon: false,
      appleStartup: false,
      windows: false,
      yandex: false,
    },
  });
  expect(images.length).toBe(1);
  expect(html).toEqual([expect.stringMatching(/<link .*png/)]);
});

test("should allow to restrict the icons to generate android icons only", async () => {
  expect.assertions(2);
  const { html, images } = await favicons(logo_png, {
    icons: {
      favicons: false,
      android: ["android-chrome-48x48.png"],
      appleIcon: false,
      appleStartup: false,
      windows: false,
      yandex: false,
    },
  });
  expect(images.length).toBe(1);
  expect(html).toEqual([
    expect.stringMatching(/webmanifest/),
    expect.stringMatching(/mobile-web-app-capable/),
    expect.stringMatching(/theme-color/),
    expect.stringMatching(/application-name/),
  ]);
});

test("should allow to restrict the icons to generate apple icons only", async () => {
  expect.assertions(2);
  const { html, images } = await favicons(logo_png, {
    icons: {
      favicons: false,
      android: false,
      appleIcon: ["apple-touch-icon-60x60.png"],
      appleStartup: false,
      windows: false,
      yandex: false,
    },
  });
  expect(images.length).toBe(1);
  expect(html).toEqual([
    expect.stringMatching(/<link .*png/),
    expect.stringMatching(/apple-mobile-web-app-capable/),
    expect.stringMatching(/apple-mobile-web-app-status-bar-style/),
    expect.stringMatching(/apple-mobile-web-app-title/),
  ]);
});

test("should allow to restrict the icons to generate and configure them simultaneously", async () => {
  const options = {
    icons: {
      favicons: false,
      android: false,
      appleStartup: false,
      windows: false,
      yandex: false,
      appleIcon: [
        {
          name: "apple-touch-icon-60x60.png",
          background: "#CCF",
        },
        "apple-touch-icon-76x76.png",
      ],
    },
  };

  const result = await favicons(logo_png, options);
  await expect(result).toMatchFaviconsSnapshot();
});
