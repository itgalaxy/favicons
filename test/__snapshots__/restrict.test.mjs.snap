// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow to restrict the icons to generate and configure them simultaneously 3`] = `
{
  "files": [],
  "html": [
    "<link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">",
    "<link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">",
    "<meta name="apple-mobile-web-app-capable" content="yes">",
    "<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">",
    "<meta name="apple-mobile-web-app-title">",
  ],
  "htmlTags": [
    FaviconHtmlTag {
      "attrs": {
        "href": "/apple-touch-icon-60x60.png",
        "rel": "apple-touch-icon",
        "sizes": "60x60",
      },
      "tag": "link",
    },
    FaviconHtmlTag {
      "attrs": {
        "href": "/apple-touch-icon-76x76.png",
        "rel": "apple-touch-icon",
        "sizes": "76x76",
      },
      "tag": "link",
    },
    FaviconHtmlTag {
      "attrs": {
        "content": "yes",
        "name": "apple-mobile-web-app-capable",
      },
      "tag": "meta",
    },
    FaviconHtmlTag {
      "attrs": {
        "content": "black-translucent",
        "name": "apple-mobile-web-app-status-bar-style",
      },
      "tag": "meta",
    },
    FaviconHtmlTag {
      "attrs": {
        "name": "apple-mobile-web-app-title",
      },
      "tag": "meta",
    },
  ],
  "images": [
    {
      "contents": null,
      "name": "apple-touch-icon-60x60.png",
    },
    {
      "contents": null,
      "name": "apple-touch-icon-76x76.png",
    },
  ],
}
`;
